app.py

```diff
@@ -1,26 +1,402 @@
+    """
+    today = datetime.now().date()
+    five_days_from_now = today + timedelta(days=5)
+    # create readable format, as should be input
+    # return [today.strftime('%Y-%m-%d'), five_days_from_now.strftime('%Y-%m-%d')]
+    return [today, five_days_from_now]
+
+
+def _filter_search_values(key: str, values: list, collection: list):
+    """
+    Filters tweet data by a single key and many values.
+
+    ## Returns a list of items which [key]-collection holds any of the given values
+
+    ### Example: 
+
+    collection holds ['#trump', '#biden', '#uselection'].  
+    values hold ['#trump', '#uselection'].
+
+    This item would be returned as it matches 2 of 3 values.
+
+    >>> list (original collection)
+    """
+    return_data = []
+    for item in collection:
+        if any(val in values for val in item[key]):
+            return_data.append(item)
+    return return_data
+
+
+def _filter_data(analyzed_tweet_data: list, start_date, end_date, hashtags, mentions, urls):
+    """
+    Filters data according to given arguments.
+    Filters by date per default.
+    Other arguments default to None and will not execute unless specifically passed.
+    """
+    # filter by dates
+    filtered_data = get_tweets_in_daterange(
+        analyzed_tweet_data, start_date, end_date)
+    print("Done filtering on date...")
+    if hashtags:
+        filtered_data = _filter_search_values(
+            'hashtags', hashtags, filtered_data)
+        print(f'Done filtering on hashtags: {hashtags}')
+    if mentions:
+        filtered_data = _filter_search_values(
+            'mentions', mentions, filtered_data)
+        print(f'Done filtering on mentions: {mentions}')
+    if urls:
+        filtered_data = _filter_search_values(
+            'tweet_urls', urls, filtered_data)
+        print(f'Done filtering on urls: {urls}')
+
+    return filtered_data
+
+#########HELPER METHODS#########
+
+
+def prepare_data(hashtags: List,
+                 tweet_amount: int,
+                 fresh_search: bool,
+                 save_plot: bool,
+                 dates: List[datetime.date],
+                 plot_type: str,
+                 search_mentions: List,
+                 search_hashtags: List,
+                 search_urls: List,
+                 remove_sentiment: str,
+                 certainty_low: float,
+                 certainty_high: float):
+    """
+    Main method that ties all the components together. Takes use of above helper methods.
+    """
+    # prepare model for analysis
     train_model_if_necessary()
+    # Verify data
+    start_date, end_date = dates
+    if (start_date > end_date):
+        raise ValueError(f'Start date {start_date} may not be later than end date {end_date}')
+    # negate bool as to get meaning worthy of get_tweets
+    # (app asks, "do you want local search?" get_tweets asks, "do you want online search?")
+    fresh_search = not fresh_search
```

Comment:
Whats this?

Reply:
Is it not clear from the documentation?

The program asks the user, *do you want to search locally or online?* (default False)

The webscraper asks the program, *do you want to search online or locally?* (default True)

We have to negate as they are opposite boolean values.

**This basically turns False to True or True to False**

Reply:
Yeah, I just meant "what does "not" do?" 
And its basically just a ! operator like we know from JavaScript. Thanks :)